### ISSUE TITLE
Check more flake8 in pre-commit

### ISSUE THREAD

**Issue by nstarman** [astropy/astropy#13735]
Right now we use a small subset of flake8 in pre-commit. Given that pre-commit runs primarily on PR diffs, it seems fine to check for more problems. There are hundreds of violations of the full flake8 checks, so I think we should sequentially add codes to pre-commit to keep correction PRs small (the PRs should either fix the problems and/or add the appropriate `noqa`).

PRs TODO
- [ ] F401 : module imported but unused
- [ ] F402 : import module from line N shadowed by loop variable
- [ ] F403 ‘from module import *’ used; unable to detect undefined names
- [ ] F404 : future import(s) name after other statements
- [ ] F405 : name may be undefined, or defined from star imports: module
- [ ] F406 : ‘from module import *’ only allowed at module level
- [ ] F407 : an undefined __future__ feature name was imported
- [ ] F501 : invalid % format literal
- [ ] F502 : % format expected mapping but got sequence
- [ ] F503 : % format expected sequence but got mapping
- [ ] F504 : % format unused named arguments
- [ ] F505 : % format missing named arguments
- [ ] F506 : % format mixed positional and named arguments
- [ ] F507 : % format mismatch of placeholder and argument count
- [ ] F508 : % format with * specifier requires a sequence
- [ ] F509 : % format with unsupported format character
- [ ] F521 : .format(...) invalid format string
- [ ] F522 : .format(...) unused named arguments
- [ ] F523 : .format(...) unused positional arguments
- [ ] F524 : .format(...) missing argument
- [ ] F525 : .format(...) mixing automatic and manual numbering
- [ ] F541 : f-string without any placeholders
- [ ] F601 : dictionary key name repeated with different values
- [ ] F602 : dictionary key variable name repeated with different values
- [ ] F621 : too many expressions in an assignment with star-unpacking
- [ ] F622 : two or more starred expressions in an assignment (a, *b, *c = d)
- [ ] F631 : assertion test is a tuple, which is always True
- [ ] F632 : use ==/!= to compare str, bytes, and int literals
- [ ] F633 : use of >> is invalid with print function
- [ ] F634 : if test is a tuple, which is always True
- [ ] F701 : a break statement outside of a while or for loop
- [ ] F702 : a continue statement outside of a while or for loop
- [ ] F703 : a continue statement in a finally block in a loop
- [ ] F704 : a yield or yield from statement outside of a function
- [ ] F706 : a return statement outside of a function/method
- [ ] F707 : an except: block as not the last exception handler
- [ ] F721 : syntax error in doctest
- [ ] F722 : syntax error in forward annotation
- [ ] F723 : syntax error in type comment
- [ ] F811 : redefinition of unused name from line N
- [ ] F821 : undefined name name
- [ ] F822 : undefined name name in __all__
- [ ] F823 : local variable name … referenced before assignment
- [ ] F831 : duplicate argument name in function definition
- [ ] F841 : local variable name is assigned to but never used
- [ ] F901 : raise NotImplemented should be raise NotImplementedError
- [ ] E999
- [ ] C901
- [ ] All the pycodestyles errors https://pycodestyle.pycqa.org/en/latest/intro.html#error-codes

**Comment by mhvk** [comment]
I'm quite fine with going for excluding checks instead of selecting them (the only ones I'd exclude are about white space around operators, which is mostly but not always good (as indeed noted in PEP8).

**Comment by nstarman** [comment]
I agree. My plan is to add checks until we've added all the ones we want, then switch to an exclusion list.
Or the reverse, start with a long exclusion list and winnow it down. Maybe the latter option is better.

**Comment by nstarman** [comment]
Hmm. flake8 in pre-commit doesn't seem to use ``--ignore``. Or at least I can't get it to work. I found https://github.com/pre-commit/pre-commit-hooks/issues/112 on the subject.

Has anyone (maybe @WilliamJamieson) used the arg ``--ignore`` in  flake8 in pre-commit?

**Comment by WilliamJamieson** [comment]
No, I just modified the, `flake8` config using `extend-ignore`, see: https://github.com/asdf-format/asdf/blob/bf80c465cd8853db434d5bf62de0cee28c7ae2bd/.flake8#L13

**Comment by nstarman** [comment]
The problem of having 2 separate configs.

**Comment by WilliamJamieson** [comment]
Sounds like we need to merge them together.

**Comment by nstarman** [comment]
The `setup.cfg` checks for everything, the pre-commit one is much more selective. I think the pre-commit one should be relaxed.

**Comment by WilliamJamieson** [comment]
Then why don't we relax the pre-commit checks to match the setup.cfg one?

**Comment by nstarman** [comment]
Apparently the ``setup.cfg`` ones are never run (?).  I've tried relaxing pre-commit and running ``pre-commit run -a`` and there are hundreds of errors

**Comment by WilliamJamieson** [comment]
Then it sounds to me like the setup.cfg ones are defunct, and should be replaced by the pre-commit one.

**Comment by mhvk** [comment]
I think the idea was similar to what pre-commit is now doing, that you'd have `flake8` run on anything you edit. Certainly, for me, `emacs` automatically looks the `setup.cfg` settings and red-wiggles everything that is not right. I think @pllim uses the same setup -- probably this is why the two of us so often have unrelated PEP8 fixes in our PRs...

Anyway, I would suggest that for pre-commit we might as well tighten it to be what is in `setup.cfg`...

**Comment by WilliamJamieson** [comment]
I suppose we could go through and start fixing all the issues the setup.cfg flake8 flags?

**Comment by eerovaher** [comment]
Wouldn't it be better to delay this until a decision about the `black` APE is made? I don't see why it would be a good idea to manually apply code style updates if `black` will do the same thing automatically soon enough.

**Comment by nstarman** [comment]
We could. However, `black` will fix a subset of these flake8 errors, the rest will still need fixing manually.

**Comment by eerovaher** [comment]
Even if manual fixes cannot be avoided entirely, wouldn't it still be less work to apply them after `black` has been adopted?

**Comment by nstarman** [comment]
Definitely would 😆.

**Comment by WilliamJamieson** [comment]
After doing some of the simpler packages, yes I agree.

**Comment by WilliamJamieson** [comment]
After doing some of the simpler packages, yes I agree.

**Comment by saimn** [comment]
Things are going a little bit too fast I think ... there are reasons for the way flake8 was configured. flake8 was run with a small subset of errors in the CI codestyle job as a compromise between people wanting a more or less strict style checking. And the config was in the CI config to allow people using a more strict style in their editor. So the relaxed check was only the codesyle CI job, but then it went in pre-commit which is used both in CI and now more frequently by developers.

Also before starting to change many lines of code for stylistic purposes I think we should think a little more about how to do this. With the isort changes and now the flake8 ones it will be more and more difficult to backport PRs without doing it manually. And is it worth doing this now when we haven't yet decided if we will use black. I mean I'm in favor of black and stricter flake8 but just a bit afraid that maintaining the release branches will be a pain.

**Comment by nstarman** [comment]
@saimn, would you agree that now that `black` will be applied, the situation is different?
We've added more than half of flake8, esp in https://github.com/astropy/astropy/pull/13853, and `black` will enforce most of the rest. I think after `black` is applied we're in a good position to enforce essentially all of flake8 (there are a few like E203 that conflict with `black` and we might choose to exclude one or two others, like E403 in `__init__`).

**Comment by nstarman** [comment]
@WilliamJamieson, when you apply the `black` PR, can you check which E/F/etc codes are automatically enforced by black?

**Comment by saimn** [comment]
@nstarman - Sure, one of the benefit of using black is that the formatting should then be pep8 compliant, except maybe a few specific cases. So I guess once black has been applied we could switch to a flake8 config where we just exclude a few rules.

**Comment by nstarman** [comment]
And maybe even switch to Ruff.... thought then we lose on flake8 plugins. Anyway, that's a discussion for after this Issue is closed.

**Comment by nstarman** [comment]
Is that we have black and ruff, this can be closed.
